### DISCUSSION TITLE
Disk and memory PE headers comparision

### DISCUSSION THREAD

**Discussion by rabbitstack** [hasherezade/pe-sieve#118] [General]
I apologize if this is a bit off-topic. I'm trying to port some of the pe-sieve implant detection techniques to [Fibratus](https://github.com/rabbitstack/fibratus). I'm already using a PE parser package that does all the heavy lifting of dissecting the PE structure from the on-disk image file. As usual, `ReadProcessMemory` is used to fetch the in-memory PE layout from the image base address. However, when it comes to comparing the PE headers, I'm hitting the wall. It turns out, only the few first fields of the DOS header are identical in both on-disk and memory PEs, while the rest of the headers differ. I know I'm missing something obvious. Could you please elaborate on how exactly pe-sieve "normalizes" the PE buffers to make possible headers comparison?
Thanks

**Comment by hasherezade** [comment]
hi @rabbitstack !
You can see details of it in the `HeadersScanner` class:
https://github.com/hasherezade/pe-sieve/blob/master/scanners/headers_scanner.cpp

**Reply by rabbitstack** [reply]
Thanks for your reply,  @hasherezade! Let me reword my initial concern. I've already explored the `headers` scanner and I'm more or less familiar with its logic. The problem I'm facing is that the remote and local module buffers differ for a typical non-implanted executable image, such as `notepad.exe`. If I compare the buffers, only the MZ and a few more fields in the DOS header are identical. Starting from the `e_lfanew` field, I can already observe they have different values in local (on-disk) and remote (memory) buffers.
I've gone through `libpeconv`, especially the `read_remote_pe` function, and I'm not sure if some sort of adjustments are made to the in-memory buffer, so the reflected DOS/NT/Section headers are equal.
For the sake of correctness, I've also inspected the on-disk and in-memory PE headers using the PEBear tool and Process Hackers. Obviously, they were different.
TIA!

**Reply by hasherezade** [reply]
@rabbitstack  - hmm, as I understand, PE-sieve doesn't detect those as modified, only your tool? Can you show me some example?

**Reply by rabbitstack** [reply]
@hasherezade Right, I suspect something is fishy in my code. I believe pe-sieve should detect those headers as unmodified. 
Let me show you some code. The [ReadArea](https://github.com/rabbitstack/fibratus/blob/70c1e1bf706d5a14ba4134b4f7f292fd8eb33b76/pkg/util/va/region.go#L226) function is essentially a port of the read remote PE logic from libpeconv. It is consumed by [ParseMem](https://github.com/rabbitstack/fibratus/blob/70c1e1bf706d5a14ba4134b4f7f292fd8eb33b76/pkg/pe/parser.go#L179) function which fetches the remote PE. The behaviour that I'm experiencing is manifested by discrepancies in the buffer read by the ParseMem and [ParseFile](https://github.com/rabbitstack/fibratus/blob/70c1e1bf706d5a14ba4134b4f7f292fd8eb33b76/pkg/pe/parser.go#L143) functions. In a nutshell, I should at least expect them to have the same number of sections. Address of entry point and base address are different, which is sort of expected due to ASLR shengaings.
Would appreciate any hints.

**Reply by hasherezade** [reply]
@rabbitstack Can you show me some executable with which this problem occurs? I would like to reproduce it and take a closer look.
The best if you could share the original, runnable exe, as well the dump created by your tool.
(You can zip them and paste here).

**Reply by rabbitstack** [reply]
I'll come up with something ASAP and share it with you.

**Reply by rabbitstack** [reply]
Hi @hasherezade,

Please find attached the binary in the https://github.com/rabbitstack/remote-pe-test repository along with the source code. The test binary will try to parse the executable image which is by default `notepad.exe`, but can get overridden with the `-exec` command line option. The same executable image is used to spawn the process from which the remote PE is read. As you can observe in the following output, reading the PE from the local (file) and remote (memory) modules, doesn't give the expected results.

```
2023/08/13 00:36:19 Trying to read local PE from C:\WINDOWS\notepad.exe executable image...
2023/08/13 00:36:19 Local PE number of sections: 7
2023/08/13 00:36:19 Local PE e_lfanew: 0xf0
2023/08/13 00:36:19 Local PE buffer: [77 90 144 0 3 0 0 0 4 0 0 0 255 255 0 0 184 0 0 0 0 0 0 0 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 240 0 0 0 14 31 186 14 0 180 9 205 33 184 1 76 205 33 84 104 105 115 32 112 114 111 103 114 97 109 32 99 97 110 110 111 116 32 98 101 32 114 117 110 32 105 110 32 68 79 83 32 109 111 100 101 46 13 13 10 36 0 0 0 0 0 0 0 134 156 132 242 194 253 234 161 194 253 234 161 194 253 234 161 203 133 121 161 156 253 234 161 137 133 238 160 200 253 234 161 137 133 233 160 198 253 234 161 137 133 239 160 221 253 234 161 137 133 235 160 197 253 234 161 194 253 235 161 166 248 234 161 137 133 226 160 222 253 234 161 137 133 23 161 195 253 234 161 137 133 21 161 195 253 234 161 137 133 232 160 195 253 234 161 82 105 99 104 194 253 234 161 0 0 0 0 0 0 0 0 80 69 0 0 100 134 7 0 180 239 152 231 0 0 0 0 0 0 0 0 240 0 34 0 11 2 14 30 0 128 2 0 0 16 3 0 0 0 0 0 160 25 0 0 0 16 0 0 0 0 0 64 1 0 0 0 0 16 0 0 0 16 0 0 10 0 0 0 10 0 0 0 10 0 0 0 0 0 0 0 0 160 5 0 0 16 0 0 112 220 5 0 2 0 96 193 0 0 8 0 0 0 0 0 0 16 1 0 0 0 0 0 0 0 16 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 0 0 0 0 9 3 0 252 3 0 0 0 160 3 0 208 225 1 0 0 112 3 0 52 20 0 0 0 0 0 0 0 0 0 0 0 144 5 0 248 2 0 0 64 228 2 0 84 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 144 151 2 0 64 1 0 0 0 0 0 0 0 0 0 0 208 152 2 0 104 11 0 0 56 4 3 0 224 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 46 116 101 120 116 0 0 0 194 123 2 0 0 16 0 0 0 128 2 0 0 16 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 0 0 96 46 114 100 97 116 97 0 0 8 166 0 0 0 144 2 0 0 176 0 0 0 144 2 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 100 97 116 97 0 0 0 192 38 0 0 0 64 3 0 0 16 0 0 0 6 ... 0]
2023/08/13 00:36:20 Launching C:\WINDOWS\notepad.exe...
2023/08/13 00:36:21 Trying to read remote PE at base address 0x7ff7c8cb0000...
2023/08/13 00:36:21 Remote PE number of sections: 6
2023/08/13 00:36:21 Remote PE e_lfanew: 0x110
2023/08/13 00:36:21 Remote PE buffer: [77 90 144 0 3 0 0 0 4 0 0 0 255 255 0 0 184 0 0 0 0 0 0 0 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 0 16 1 0 0 14 31 186 14 0 180 9 205 33 184 1 76 205 33 84 104 105 115 32 112 114 111 103 114 97 109 32 99 97 110 110 111 116 32 98 101 3
2 114 117 110 32 105 110 32 68 79 83 32 109 111 100 101 46 13 13 10 36 0 0 0 0 0 0 0 59 192 0 123 127 161 110 40 127 161 110 40 127 161 110 40 192 221 1
06 41 117 161 110 40 192 221 109 41 123 161 110 40 192 221 107 41 83 161 110 40 192 221 111 41 121 161 110 40 52 217 106 41 125 161 110 40 52 217 104 41
 126 161 110 40 118 217 253 40 77 161 110 40 52 217 111 41 118 161 110 40 127 161 111 40 98 163 110 40 166 220 103 41 77 161 110 40 166 220 145 40 126 1
61 110 40 127 161 249 40 126 161 110 40 166 220 108 41 126 161 110 40 82 105 99 104 127 161 110 40 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 80 69 0 0 100 134 6 0
 188 17 174 100 0 0 0 0 0 0 0 0 240 0 34 0 11 2 14 35 0 134 8 0 0 218 6 0 0 0 0 0 108 45 8 0 0 16 0 0 0 0 203 200 247 127 0 0 0 16 0 0 0 2 0 0 6 0 0 0 0
 0 0 0 6 0 0 0 0 0 0 0 0 160 15 0 0 4 0 0 0 0 0 0 2 0 96 129 0 0 16 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 16 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 0 0 16 0 0 0 0 0 0 
0 0 0 0 0 120 222 12 0 128 2 0 0 0 176 14 0 48 217 0 0 0 64 14 0 60 99 0 0 0 0 0 0 0 0 0 0 0 144 15 0 4 13 0 0 224 53 11 0 84 0 0 0 0 0 0 0 0 0 0 0 0 0 
0 0 0 0 0 0 128 54 11 0 40 0 0 0 160 52 11 0 64 1 0 0 0 0 0 0 0 0 0 0 0 160 8 0 88 16 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 46 116 101 120
 116 0 0 0 103 132 8 0 0 16 0 0 0 134 8 0 0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 0 0 96 46 114 100 97 116 97 0 0 10 129 4 0 0 160 8 0 0 130 4 0 0 138 8 0 0 
0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 100 97 116 97 0 0 0 176 10 1 0 0 48 13 0 0 246 0 0 0 12 13 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 192 46 112 100 97 116 97 
0 0 60 99 0 0 0 64 14 0 0 100 0 0 0 2 14 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 114 115 114 99 0 0 0 48 217 0 0 0 176 14 0 0 218 0 0 0 102 14 0 0 0 0 0 
0 0 0 0 0 0 0 0 64 0 0 64 46 114 101 108 111 99 0 0 4 13 0 0 0 144 15 0 0 14 0 0 0 64 15 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 66 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
```

**Reply by rabbitstack** [reply]
Hi @hasherezade 

I'm not intending to rush, but just want to check you saw this message and that you have all the bits needed for your investigation.

Thanks

**Reply by hasherezade** [reply]
@rabbitstack - sorry, I was busy with other things lately, I am gonna check it soon...

**Reply by hasherezade** [reply]
@rabbitstack - could you please run my modified snippet (https://gist.github.com/hasherezade/df150d35cf21f67487f6fc87da9729d5)  and send me the output?
I will need the files: `local_header.hex` and `remote_header.hex`.

I tested it on my machine but I couldn't reproduce the same issues that you've got. 
My output from your binary:
```
C:\Users\tester\Desktop\remote-pe-test>pe-test.exe
2023/08/28 06:33:22 Trying to read local PE from C:\Windows\notepad.exe executable image...
2023/08/28 06:33:22 Local PE number of sections: 7
2023/08/28 06:33:22 Local PE e_lfanew: 0xf8
2023/08/28 06:33:22 Local PE buffer: [77 90 144 0 3 0 0 0 4 0 0 0 255 255 0 0 184 0 0 0 0 0 0 0 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 248 0 0 0 14 31 186 14 0 180 9 205 33 184 1 76 205 33 84 104 105 115 32 112 114 111 103 114 97 109 32 99 97 110 110 111 116 32 98 101 32 114 117 110 32 105 110 32 68 79 83 32 109 111 100 101 46 13 13 10 36 0 0 0 0 0 0 0 101 57 215 116 33 88 185 39 33 88 185 39 33 88 185 39 40 32 42 39 17 88 185 39 53 51 189 38 43 88 185 39 53 51 186 38 34 88 185 39 53 51 184 38 40 88 185 39 33 88 184 39 11 93 185 39 53 51 177 38 63 88 185 39 53 51 188 38 62 88 185 39 53 51 68 39 32 88 185 39 53 51 70 39 32 88 185 39 53 51 187 38 32 88 185 39 82 105 99 104 33 88 185 39 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 80 69 0 0 100 134 7 0 24 54 166 59 0 0 0 0 0 0 0 0 240 0 34 0 11 2 14 20 0 94 2 0 0 230 0 0 0 0 0 0 16 84 2 0 0 16 0 0 0 0 0 64 1 0 0 0 0 16 0 0 0 2 0 0 10 0 0 0 10 0 0 0 10 0 0 0 0 0 0 0 0 144 3 0 0 4 0 0 211 241 3 0 2 0 96 193 0 0 8 0 0 0 0 0 0 16 1 0 0 0 0 0 0 0 16 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 0 0 0 216 230 2 0 68 2 0 0 0 112 3 0 216 11 0 0 0 64 3 0 136 17 0 0 0 0 0 0 0 0 0 0 0 128 3 0 232 2 0 0 160 188 2 0 84 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 16 119 2 0 24 1 0 0 0 0 0 0 0 0 0 0 40 120 2 0 16 9 0 0 240 223 2 0 224 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 46 116 101 120 116 0 0 0 207 92 2 0 0 16 0 0 0 94 2 0 0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 0 0 96 46 114 100 97 116 97 0 0 214 152 0 0 0 112 2 0 0 154 0 0 0 98 2 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 100 97 116 97 0 0 0 136 39 0 0 0 16 3 0 0 14 0 0 0 252 2 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 192 46 112 100 97 116 97 0 0 136 17 0 0 0 64 3 0 0 18 0 0 0 10 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 100 105 100 97 116 0 0 120 1 0 0 0 96 3 0 0 2 0 0 0 28 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 192 46 114 115 114 99 0 0 0 216 11 0 0 0 112 3 0 0 12 0 0 0 30 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 114 101 108 111 99 0 0 232 2 0 0 0 128 3 0 0 4 0 0 0 42 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 66 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 204 204 204 204 204 204 204 204 64 85 72 141 108 36 225 72 129 236 208 0 0 0 72 139 5 131 4 3 0 72 51 196 72 137 69 15 72 139 69 127 69 51 201 72 137 69 255 69]
2023/08/28 06:33:22 Launching C:\Windows\notepad.exe...
2023/08/28 06:33:23 Trying to read remote PE at base address 0x7ff723760000...
2023/08/28 06:33:23 Remote PE number of sections: 7
2023/08/28 06:33:23 Remote PE e_lfanew: 0xf8
2023/08/28 06:33:23 Remote PE buffer: [77 90 144 0 3 0 0 0 4 0 0 0 255 255 0 0 184 0 0 0 0 0 0 0 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 248 0 0 0 14 31 186 14 0 180 9 205 33 184 1 76 205 33 84 104 105 115 32 112 114 111 103 114 97 109 32 99 97 110 110 111 116 32 98 101 32 114 117 110 32 105 110 32 68 79 83 32 109 111 100 101 46 13 13 10 36 0 0 0 0 0 0 0 101 57 215 116 33 88 185 39 33 88 185 39 33 88 185 39 40 32 42 39 17 88 185 39 53 51 189 38 43 88 185 39 53 51 186 38 34 88 185 39 53 51 184 38 40 88 185 39 33 88 184 39 11 93 185 39 53 51 177 38 63 88 185 39 53 51 188 38 62 88 185 39 53 51 68 39 32 88 185 39 53 51 70 39 32 88 185 39 53 51 187 38 32 88 185 39 82 105 99 104 33 88 185 39 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 80 69 0 0 100 134 7 0 24 54 166 59 0 0 0 0 0 0 0 0 240 0 34 0 11 2 14 20 0 94 2 0 0 230 0 0 0 0 0 0 16 84 2 0 0 16 0 0 0 0 118 35 247 127 0 0 0 16 0 0 0 2 0 0 10 0 0 0 10 0 0 0 10 0 0 0 0 0 0 0 0 144 3 0 0 4 0 0 211 241 3 0 2 0 96 193 0 0 8 0 0 0 0 0 0 16 1 0 0 0 0 0 0 0 16 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 0 0 16 0 0 0 0 0 0 0 0 0 0 0 216 230 2 0 68 2 0 0 0 112 3 0 216 11 0 0 0 64 3 0 136 17 0 0 0 0 0 0 0 0 0 0 0 128 3 0 232 2 0 0 160 188 2 0 84 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 16 119 2 0 24 1 0 0 0 0 0 0 0 0 0 0 40 120 2 0 16 9 0 0 240 223 2 0 224 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 46 116 101 120 116 0 0 0 207 92 2 0 0 16 0 0 0 94 2 0 0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 0 0 96 46 114 100 97 116 97 0 0 214 152 0 0 0 112 2 0 0 154 0 0 0 98 2 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 100 97 116 97 0 0 0 136 39 0 0 0 16 3 0 0 14 0 0 0 252 2 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 192 46 112 100 97 116 97 0 0 136 17 0 0 0 64 3 0 0 18 0 0 0 10 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 100 105 100 97 116 0 0 120 1 0 0 0 96 3 0 0 2 0 0 0 28 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 192 46 114 115 114 99 0 0 0 216 11 0 0 0 112 3 0 0 12 0 0 0 30 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 64 46 114 101 108 111 99 0 0 232 2 0 0 0 128 3 0 0 4 0 0 0 42 3 0 0 0 0 0 0 0 0 0 0 0 0 0 64 0 0 66 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
```

The only modified element of the header was ImageBase, which makes sense:

![img_base_diff](https://github.com/hasherezade/pe-sieve/assets/3115348/ab2ff0b3-1250-4b7c-821f-e25617251921)

I tried to convert and compare your dumped output, but it seems like some bytes in between are missing (left: local header, right:remote header):

![compared](https://github.com/hasherezade/pe-sieve/assets/3115348/3b665369-7201-4528-9a12-bcec6b0e5f0b)

The Rich Header seems the same, so it is not like a wrong file was used for the comparison. But rather some problem with reading the remote memory. Anyways, I will prefer to see the content dumped into a file, just to doublecheck that this is not a problem coming from how the buffer was printed.

**Reply by rabbitstack** [reply]
Hi @hasherezade ,

I've just pushed those files to https://github.com/rabbitstack/remote-pe-test. Let me know if you need anything more from me.

Regards

**Reply by rabbitstack** [reply]
Just wanted to bring up one important fact. I can vaguely recall I didn't have local/remote PE header discrepancies before I upgraded from Windows 10 to 11. So, it is either the OS, the Notepad image, or both of them : ).

**Reply by hasherezade** [reply]
thank you, I checked it. So, I was wrong when I said before "The Rich Header seems the same, so it is not like a wrong file was used for the comparison. ". In fact the wrong file WAS used for the comparison. Probably a different version of the Notepad was loaded in memory, and different was read from the disk.
You can have a look at those dumped headers in PE-bear. For example, the Rich Header of both is fully different:

### Remote:

![remote2](https://github.com/hasherezade/pe-sieve/assets/3115348/57c7014b-028e-47f0-843a-bc32e5bb0bc0)

### Local:

![local1](https://github.com/hasherezade/pe-sieve/assets/3115348/b95f892d-5e36-43eb-932f-ec7c206aa7d5)

Did you scan exactly the same instance of Notepad.exe (the one that was spawned by your application) with PE-sieve? Is it not detecting it as modified?
Please show me what exactly PE-sieve prints on the screen when it scans this process...

**Reply by hasherezade** [reply]
@rabbitstack - did you observe the same behavior in case of other applications too? which ones are affected?
Unfortunately I don't have an access to a machine with Windows 11 right now, so I cannot test it...
But hopefully seeing the output from PE-sieve would give me some more hints. I need to see all the paths that it scanned. 
By seeing the headers I am starting to suspect that the image path is different here than the path of the actual mapped executable.

**Reply by rabbitstack** [reply]
Hi @hasherezade ,

I believe I've found the root cause for this funny behavior. Your assumption was correct when you stated a different image could end up mapping into the process address space. So, basically, the local PE was parsed from the `C:\Windows\notepad.exe`, while spinning up the process for this image would in reality load the `C:\Program Files\WindowsApps\Microsoft.WindowsNotepad_11.2306.15.0_x64__8wekyb3d8bbwe\Notepad\Notepad.exe` executable image. 

Thank you for your time and patience.

Kind regards.

**Reply by hasherezade** [reply]
hi @rabbitstack !
I am glad it got solved! Yes, sometimes Windows does this implicit path redirection. Another thing you need to be careful about it [FS redirection between System32 and SysWow64 paths](https://learn.microsoft.com/en-us/windows/win32/winprog64/file-system-redirector). If not disabled, it causes an implicit conversion of System32 paths to SysWow64 when the path is requested by a 32-bit application. So you may again read a different version of the same executable. Just something to keep in mind.

If everything is clarified now, feel free to mark this question as answered.

**Reply by rabbitstack** [reply]
Thanks for the clarification and the heads-up about path redirection. I don't seem to be able to mark the reply as answered. Apparently, only discussions pertaining to the Q/A category can be marked as answered. I'm willing to close this discussion as resolved if this sounds right to you.

**Reply by hasherezade** [reply]
yeah, closing the discussion is fine :)
